mwmbl/tinysearchengine/indexer.py

```diff
@@ -65,19 +65,39 @@ def from_bytes(data: bytes):
         values = json.loads(data[constant_length:].decode('utf8'))
         return TinyIndexMetadata(**values)
 
+# Find the optimal amount of data that fits onto a page
+# We do this by leveraging binary search to quickly find the index where:
+#     - index+1 cannot fit onto a page
+#     - <=index can fit on a page
+def _binary_search_fitting_size(compressor: ZstdCompressor, page_size: int, items:list[T], lo:int, hi:int):
+    # Base case: our binary search has gone too far
+    if lo > hi:
+        return -1, None
+    # Check the midpoint to see if it will fit onto a page
+    mid = (lo+hi)//2
+    compressed_data = compressor.compress(json.dumps(items[:mid]).encode('utf8'))
+    size = len(compressed_data)
+    if size > page_size:
+        # We cannot fit this much data into a page
+        # Reduce the hi boundary, and try again
+        return _binary_search_fitting_size(compressor,page_size,items,lo,mid-1)
```

Comment:
Style: there should be a space after the commas.

Side note: for future work, we could use a tool called `black` to format our code. Then we could have a check in GitHub that the PR satisfies black formatting.

Reply:
Great point, we should definitely do that. I will look into it.
Addressed the formatting manually for this change